_posts/2017-04-24-solarized.md

```diff
@@ -0,0 +1,545 @@
+a solarized look-and-feel. Google uses around 80 parameters for its design, and
+will only pick some colors from GTK, while using custom colors for other
+parameters.
+In order to truly customize the Chrome theme, it's require to build an
+extension-like theme, which can then be uploaded to the Chrome store and used
+from there.
+I used this [Solarized Dark
+theme](https://chrome.google.com/webstore/detail/solarized-dark/kedemblecjmofcmppbecaagebmokigml), and since I couldn't find a Solarized Light version, I [built it myself](https://github.com/Fandekasp/google-chrome-solarized-light)
+
+I couldn't figure out any way to reload Chrome the same way it reloads itself when
+updating the settings or installing a new theme from the Chrome store.
+Therefore, I went with a restart approach, specifying the
+`--load-extension=/path/to/theme` parameter to the google google-chrome
+executable. Another problem naturally occured from that decision: When
+restarting an application, i3 will detect the windows being closed and
+automatically re-tile the other windows, then repeat the process once chrome is
+restarted, potentially breaking the window layout we had defined prior to
+changing the theme. To paliate this issue, we can improve the script to save the
+current layout before closing Chrome, then restore the layout once Chrome is
+restarted.
+
+```bash
+if [[ $(ps auxw|grep chrome|grep -v grep|wc -l) -gt 0 ]]; then
+    mapfile -t workspaces <<< $(i3-msg -t get_workspaces | jq -r '.[] | .name')
+
+    # Add save-tree logic here
+    for ws in "${workspaces[@]}"; do
+        i3-save-tree --workspace "$ws" > ~/.config/i3sessions/workspace-${ws:0:1}.json
+        # Swallows all
+        perl -e "s/\/\/ \"/\"/g" -pi ~/.config/i3sessions/workspace-${ws:0:1}.json
+    done
+
+    # Kill Chrome
+    killall chrome
+
+    # Swallow other windows
+    ids=$(xdotool search --onlyvisible --name '.*')
+    IFS='
+    '
+    ids=( $ids )
+    for id in "${ids[@]}"; do
+        xdotool windowunmap $id
+    done
+
+    # Restore logic
+    for ws in "${workspaces[@]}"; do
+        i3-msg "workspace $ws; append_layout ~/.config/i3sessions/workspace-${ws:0:1}.json"
+    done
+
+    # Go back to first workspace
+    i3-msg "workspace ${workspaces[0]}"
+
+    # Restart Chrome
+    exec google-chrome-stable --load-extension=${HOME}/Projects/personal/google-chrome-solarized-${newmode}-theme &
+
+    # Give some time for i3 to append the layout and chrome to start
+    sleep 0.5
+
+    # Unswallow other windows
+    for id in "${ids[@]}"; do
+        xdotool windowmap $id
+    done
+fi
+```
+While saving the layout with `i3-save-tree` is straightforward, restoring it
+with `i3-msg "append_layout ..."` was quite difficult, because it will simply
+add new containers next to existing containers. The trick was to use `xdotool`
+to swallow existing windows like Termite, and unswallow them for i3 to add them
```

Comment:
Added "To clarify the terminology here, "swallowing" is similar to sending a window to
the scratchpad, or in other words hide the window in a virtual workspace, which
is managed by xdotool."